driver.go

```diff
@@ -64,213 +62,415 @@ func (d *Driver) Close() {
 }
 
-func (s *Session) Close() {
-	s.Session.Close()
-}
+// // LoadNode attempts to load a node with the given ID. The ID is
+// // looked up in id node property
+// func (s *Session) LoadNode(tx neo4j.Transaction, ID string) (ls.Node, int64, error) {
+// 	ret, err := tx.Run(fmt.Sprintf("MATCH (node {%s:$id}) RETURN node", PropNodeID), map[string]interface{}{"id": ID})
+// 	if err != nil {
+// 		return nil, 0, err
+// 	}
+// 	if ret.Next() {
+// 		record := ret.Record()
+// 		if ret.Next() {
+// 			return nil, 0, ErrMultipleFound(ID)
+// 		}
+// 		node, ok := record.Values[0].(neo4j.Node)
+// 		if !ok {
+// 			return nil, 0, ls.ErrNotFound(ID)
+// 		}
+// 		return s.MakeNode(node), node.Id, nil
+// 	}
+// 	return nil, 0, ls.ErrNotFound(ID)
+// }
 
-// LoadNode attempts to load a node with the given ID. The ID is
-// looked up in id node property
-func (s *Session) LoadNode(tx neo4j.Transaction, labels []string, ID string) (ls.Node, int64, error) {
-	query := fmt.Sprintf("MATCH (node%s {%s:$id}) RETURN node", makeLabels(labels), PropNodeID)
-	data := map[string]interface{}{"id": ID}
-	s.Logf("Query: %s with %v", query, data)
-	ret, err := tx.Run(query, data)
-	if err != nil {
-		return nil, 0, err
+// // MakeNode builds a graph node from the given db node
+// func (s *Session) MakeNode(node neo4j.Node) ls.Node {
+// 	retNode := ls.NewNode(fmt.Sprint(node.Props[PropNodeID]), node.Labels...)
+// 	if x, ok := node.Props[PropNodeValue]; ok {
+// 		if str, ok := x.(string); ok {
+// 			retNode.SetValue(str)
+// 		}
+// 	}
+// 	buildPropertyMap(node.Props, retNode.GetProperties())
+// 	return retNode
+// }
+
+// func buildPropertyMap(in map[string]interface{}, out map[string]*ls.PropertyValue) {
+// 	for k, v := range in {
+// 		switch k {
+// 		case PropNodeValue:
+// 		case PropNodeID:
+// 		default:
+// 			switch val := v.(type) {
+// 			case []interface{}:
+// 				arr := make([]string, 0, len(val))
+// 				for _, x := range val {
+// 					arr = append(arr, fmt.Sprint(x))
+// 				}
+// 				out[k] = ls.StringSlicePropertyValue(arr)
+// 			default:
+// 				out[k] = ls.StringPropertyValue(fmt.Sprint(val))
+// 			}
+// 		}
+// 	}
+// }
+
+// CreateGraph creates a graph and returns the neo4j ID of the root node
+func CreateGraph(session *Session, tx neo4j.Transaction, nodes []graph.Node) (int64, error) {
+	nodeIds := make(map[graph.Node]int64)
+	allNodes := make(map[graph.Node]struct{})
+	entityNodes := make(map[graph.Node]struct{})
+	// Find triples:
+	for _, node := range nodes {
+		allNodes[node] = struct{}{}
 	}
-	if ret.Next() {
-		record := ret.Record()
-		if ret.Next() {
-			return nil, 0, ErrMultipleFound(ID)
+	for node := range allNodes {
+		hasEdges := false
+		var root graph.Node
+		var en []graph.Node
+		if _, exists := node.GetProperty(ls.EntitySchemaTerm); exists {
+			root = node
+			en = ls.GetEntityIDNodes(root)
+			for _, n := range en {
+				entityNodes[n] = struct{}{}
+			}
+		}
+		if id := session.existsDB(tx, node); id != -1 {
```

Comment:
GetEntityIDNodes()? Does this exist?

If a node is root, then it has EntitySchemaTerm property. It should also have an EntityIDProperty. This gives the ID you can use to look it up in the DB

So, if a node is root, get the EntityID value from that root node. Look it up in the db, and get its DB node. The EntityID can be a string, or a []string. If it is a []string, you have to look it up as a slice like that. Your test graphs should contain entityId values as well.

Once you find a root that is in the DB, use IterateDescendants with the ls.FollowEdgesInEntity edgeFunc to collect all nodes in the entity. These are the nodes that are already in db, so remove them from the nodes array, they will not be inserted.